xla/service/gpu/gpu_windowed_einsum_handler.cc

```diff
@@ -142,6 +143,134 @@ absl::StatusOr<bool> HandleAgWindowedEinsumLoop(HloComputation* comp,
   return changed;
 }
 
+class WindowedEinsumVisitor : public DfsHloRewriteVisitor {
+ public:
+  explicit WindowedEinsumVisitor(
+      std::vector<GpuWindowedEinsumHandler::WindowedEinsumAgLoops>&
+          all_ag_loops)
+      : all_ag_loops_(all_ag_loops) {}
+  // Rewrites a allgather-dot pattern that shares the same operand
+  // with a windowed einsum loop to consume the output of the loop
+  // and remove the all-gather.
+  absl::Status HandleDot(HloInstruction* dot) override {
+    CHECK_EQ(dot->opcode(), HloOpcode::kDot);
+    for (GpuWindowedEinsumHandler::WindowedEinsumAgLoops ag_loop :
+         all_ag_loops_) {
+      HloInstruction* loop = ag_loop.loop;
+      HloInstruction* lhs_ag = nullptr;
+      HloInstruction* rhs_ag = nullptr;
+
+      if (Match(dot, m::Dot(m::AllGather(&lhs_ag), m::Op())) ||
```

Comment:
Is it possible to reduce LHS/RHS code duplication?

Reply:
updated